### PR TITLE
Use float to string kernel

### LINKED ISSUE #4204: casting double to string does not match Spark
**Describe the bug**
I tried to cast `5.0e-10` to string. On Spark 3.2, I got `"5.0E-10"`; on spark-rapids I got `"5.0e-10"`

**Steps/Code to reproduce bug**
Spark result:
```
scala> val schema = StructType(Array(StructField("a", DoubleType)))
schema: org.apache.spark.sql.types.StructType = StructType(StructField(a,DoubleType,true))

scala> val df0 = spark.createDataFrame(spark.sparkContext.parallelize(data0), schema)
<console>:29: error: not found: value data0
       val df0 = spark.createDataFrame(spark.sparkContext.parallelize(data0), schema)
                                                                      ^

scala> val data0 = Seq(Row(5e-10))
data0: Seq[org.apache.spark.sql.Row] = List([5.0E-10])

scala> val df0 = spark.createDataFrame(spark.sparkContext.parallelize(data0), schema)
df0: org.apache.spark.sql.DataFrame = [a: double]

scala> df0.show
+-------+
|      a|
+-------+
|5.0E-10|
+-------+
```
spark-rapids result:
```
scala> val schema = StructType(Array(StructField("a", DoubleType)))
schema: org.apache.spark.sql.types.StructType = StructType(StructField(a,DoubleType,true))

scala> val data0 = Seq(Row(5e-10))
data0: Seq[org.apache.spark.sql.Row] = List([5.0E-10])

scala> val df0 = spark.createDataFrame(spark.sparkContext.parallelize(data0), schema)
df0: org.apache.spark.sql.DataFrame = [a: double]

scala> df0.sqlContext.setConf("spark.rapids.sql.castFloatToString.enabled", "true")

scala> df0.show
21/11/24 16:29:40 WARN GpuOverrides: 
!Exec <CollectLimitExec> cannot run on GPU because the Exec CollectLimitExec has been disabled, and is disabled by default because Collect Limit replacement can be slower on the GPU, if huge number of rows in a batch it could help by limiting the number of rows transferred from GPU to CPU. Set spark.rapids.sql.exec.CollectLimitExec to true if you wish to enable it
  @Partitioning <SinglePartition$> could run on GPU
  *Exec <ProjectExec> will run on GPU
    *Expression <Alias> cast(a#1 as string) AS a#4 will run on GPU
      *Expression <Cast> cast(a#1 as string) will run on GPU
    !NOT_FOUND <RDDScanExec> cannot run on GPU because no GPU enabled version of operator class org.apache.spark.sql.execution.RDDScanExec could be found
      @Expression <AttributeReference> a#1 could run on GPU

+-------+                                                                       
|      a|
+-------+
|5.0e-10|
+-------+

```

**Expected behavior**
I hope in rapids, it gives `5.0E-10`

**Environment details (please complete the following information)**
Spark 3.2.0
rapids 22.02.0
cudf 22.02.0
using spark-shell on my desktop
`setConf("spark.rapids.sql.castFloatToString.enabled", "true")`

**Additional context**
This issue is related to https://github.com/NVIDIA/spark-rapids/issues/4028.

### PR DESCRIPTION
**Pull request by thirtiseven**
This PR updates the float to string casting to a new kernel in JNI. It won't match Spark's behavior exactly, but the results are closer than the current version.

The JNI kernel is part of the format_number support, so I split it out as a subtask. 

This PR uses [Ryū: fast float-to-string conversion (PLDI'18)](https://github.com/ulfjack/ryu) as the solution for casting float/double to string. The results differ from the output of Spark's in some cases: sometimes the output is shorter (which is arguably more accurate) and sometimes the output may differ in the precise digits output (e.g., see https://github.com/ulfjack/ryu/issues/83).

In most cases, the result will match Spark's results, and in the cases where it does not, the values will match when we cast them back to float. 

Depends on: https://github.com/NVIDIA/spark-rapids-jni/pull/1508

There is some related discussion in format_number issue #9173. 

Closes #4204

**performance test results**

50000000 random number generated by BigDataGen:

test code:

```scala
spark.time(df.selectExpr("COUNT(cast(a as string)) as a", "COUNT(cast(a as string)) as b", "COUNT(cast(a as string)) as c", "COUNT(cast(a as string)) as d", "COUNT(cast(a as string)) as e").show())

spark.time(df.selectExpr("COUNT(cast(b as string)) as a", "COUNT(cast(b as string)) as b", "COUNT(cast(b as string)) as c", "COUNT(cast(b as string)) as d", "COUNT(cast(b as string)) as e").show())

```

| Data Type | GPU Time (ms) | CPU Time (ms) | Speed up |
|-|-|-|-|
| float | 282 | 5,221 | 18.51x |
| double | 379 | 14,638 |  38.62x |

### CODE REVIEW

**Comment by thirtiseven** [comment]
build

**Comment by thirtiseven** [comment]
build

**Comment by revans2** [comment]
I filed a PR to fix the markdown failure https://github.com/NVIDIA/spark-rapids/issues/10029

File: docs/compatibility.md [hunk 1]
```diff
@@ -711,9 +711,8 @@ to `false`.
 
 ### Float to String
 
-The GPU will use different precision than Java's toString method when converting floating-point data
-types to strings. The GPU uses a lowercase `e` prefix for an exponent while Spark uses uppercase
-`E`. As a result the computed string can differ from the default behavior in Spark.
+The GPU use [ryu](https://github.com/ulfjack/ryu) as the solution when converting floating-point data
```
Comment (revans2) [inline docs/compatibility.md:714]: nit: The Rapids Accelerator for Apache Spark uses a method based on [ryu](https://github.com/ulfjack/ryu) when converting floating point data type to string. ...

File: integration_tests/src/main/python/cast_test.py [hunk 2]
```diff
@@ -304,7 +304,22 @@ def test_cast_array_to_string(data_gen, legacy):
     _assert_cast_to_string_equal(
         data_gen,
         {"spark.sql.legacy.castComplexTypesToString.enabled": legacy})
-
+    
+def test_cast_float_to_string():
```
Comment (revans2) [inline integration_tests/src/main/python/cast_test.py:308]: This does not need to be an approximate float?  How many DATAGEN_SEEDS have you tested with this?

File: integration_tests/src/main/python/cast_test.py [hunk 3]
```diff
@@ -304,7 +304,22 @@ def test_cast_array_to_string(data_gen, legacy):
...
+
+def test_cast_double_to_string():
+    conf = {"spark.rapids.sql.castFloatToString.enabled": True}
+    cast_func = lambda spark: unary_op_df(spark, DoubleGen()).selectExpr("cast(a as string)").collect()
+    from_cpu = with_cpu_session(cast_func, conf)
+    from_gpu = with_gpu_session(cast_func, conf)
+    cast_to_float_func = lambda row: row.a if row.a is None or row.a == 'NaN' else float(row.a)
+    from_cpu_float = list(map(cast_to_float_func, from_cpu))
+    from_gpu_float = list(map(cast_to_float_func, from_gpu))
+    assert from_cpu_float == from_gpu_float
```
Comment (revans2) [inline integration_tests/src/main/python/cast_test.py:322]: But NaN != NaN, so how will from_cpu_float ever be equal to from_gpu_float if there is a NaN in them?

Comment (revans2) [reply integration_tests/src/main/python/cast_test.py:322]: Okay so we are mixing types within an array. That is rather confusing but I guess it is okay.

File: integration_tests/src/main/python/cast_test.py (hunk 2 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/cast_test.py:308]: I tested 10 more times with length=200000 and they all passed. 

The difference between ryu and jdk's float to string is that jdk sometimes keeps more precision, but ryu keeps it as short as possible. The results will be the same float when converted back. So I think it makes sense that we don't need an approximate float here.

File: docs/compatibility.md (hunk 1 above)
Comment (thirtiseven) [reply docs/compatibility.md]: done.

File: integration_tests/src/main/python/cast_test.py (hunk 2 above)
Comment (thirtiseven) [reply integration_tests/src/main/python/cast_test.py:308]: And also the float to string is fully matched to cpu but double to string is not. It is the reason to test float to string and double to string in two different ways.

Seems the compatibility doc for float to string is outdated, so I'm not sure if we are aware that double to string is not fully matched with cpu. Filed #10037

**Review by revans2 — APPROVED** [review]

**Comment by thirtiseven** [comment]
build

**Pull request merged**